README.md
# SURF: Some Useful R Functions
This repository hosts development of the R package `SURF`.

- Licensing information is given in the [LICENSE][] file.
- Built tarballs of the `SURF` package are available in the [builds][] 
  directory.
- Stand-alone documentation is available in the [documentation][docs] directory.
- The source files for the most recent stable version of `mibrr` are available 
  in the [source][src] directory.
  
`SURF` is alpha software created largely for my own personal use (i.e., so that 
I don't have to keep re-implementing these functions). If you choose to use 
`SURF`, please expect frequent---and dramatic---changes to the package's 
functionality and user interface. Please report any bugs that you encounter in 
the issues section of the project page.

## Installation
The best way to install `SURF` is to use the `devtools::install_github` 
function.

1. First, make sure that you have `devtools` installed on your system
2. Next, execute the following lines:

		library(devtools)
		install_github("kylelang/SURF/source/SURF")
    
3. Finally, load `SURF` and enjoy:

		library(SURF)

If the `devtools`-based approach does not work, you can download one of the
built tar-balls from the [builds][] directory and manually install the package
from source by executing the following lines:

	install.packages(pkgs  = "/SAVE_PATH/SURF_VERSION.tar.gz",
	                 repos = NULL,
                     type  = "source")

Where *SAVE_PATH* is replaced by the (relative or absolute) file path to the
location where you saved the tar-ball, and *VERSION* is replaced with the correct
version number for the tar-ball that you downloaded.

[builds]:  https://github.com/kylelang/SURF/tree/master/builds/
[docs]:    https://github.com/kylelang/SURF/tree/master/documentation/
[src]:     https://github.com/kylelang/SURF/tree/master/source/SURF
[LICENSE]: https://github.com/kylelang/SURF/blob/master/LICENSE


CHANGELOG.md
# Change Log
All notable changes to the **SURF** project will be documented in this file.

The format is based on [Keep a Changelog][kacl], and this project adheres to
[Semantic Versioning][sv].

## 0.0.0.9004 - 2018-10-15 - ACTIVE

### Added
- A new function, `simCovData`, to simulate data according to a basic covariance 
  structure.

### Changed
- Improved implementation of `simRegData` and `imposeMissData` functions.
- The `nPreds` argument in `simRegData` is now `nVars`.

## 0.0.0.9003 - 2018-06-01

### Changed
- Changed the `collin` argument to `sigma` in `simRegData`.
- Added the option to specify the entire covariance matrix of the predictors in 
  `simRegData`.
- Improved the implementations of the `surfL` and `surfW` functions.
  
## 0.0.0.9002 - 2017-11-17

### Added
- `calcMode` now estimates the mode for continous variables when `discrete = 
  FALSE`
- Package startup message

### Changed
- When simulating MAR missing data with `imposeMissData` the linear predictor is 
  now constructed using available cases

    - Allows for incomplete MAR predictor matrices
	- Throws an error if any rows are completely missing

## 0.0.0.9001 - 2017-11-09

### Added
- new function `rangeNorm` and associated doc page
- new function `calcMode` and associated doc page
- new function `f2n` and associated doc page
- doc file for `testImps.RData` example data

### Change
- included licensing header in all source files

## 0.0.0.9000 - 2017-11-03

### Added
- Initialized package
- Populated initial set of functions:

    1. `simRegData`
	1. `imposeMissData`
	1. `plotImps`
	
[kacl]: http://keepachangelog.com/
[sv]:   http://semver.org/
[hw]:   http://r-pkgs.had.co.nz/



testing/testPackage.R
### Title:    Test the SURF Package
### Author:   Kyle M. Lang
### Created:  2017-NOV-03
### Modified: 2018-JUN-01

library(SURF)

### TEST HELP EXAMPLES ###

### simRegData ###

testData <- simRegData(nObs  = 50000,
                       nVars = 10,
                       r2    = 0.5,
                       sigma = 0.2,
                       beta  = matrix(c(0.25, rep(0.75, 10)))
                       )

summary(lm(testData[ , 1] ~ as.matrix(testData[ , -1])))
cor(testData[ , -1])

sigma       <- matrix(0.35, 10, 10)
diag(sigma) <- 1.0

testData <- simRegData(nObs  = 50000,
                       r2    = 0.5,
                       sigma = sigma,
                       beta  = matrix(c(0.25, rep(0.75, 10)))
                       )

summary(lm(testData[ , 1] ~ as.matrix(testData[ , -1])))
cor(testData[ , -1])

### simCovData ###
testData <- simCovData(nObs = 50000, sigma = 0.5, nVars = 5, scales = 1.5)
cor(testData)
cov(testData)
sqrt(diag(cov(testData)))

head(testData)

### imposeMissData ###

testData <- simRegData(nObs  = 500,
                       nVars = 10,
                       r2    = 0.5,
                       sigma = 0.2,
                       beta  = matrix(c(0.25, rep(0.75, 10)))
                       )

missData <- imposeMissData(data     = testData,
                           targets  = list(mar  = c("y", "x1"),
                                           mcar = c("x2", "x3"),
                                           mnar = c("x4", "x5")
                                           ),
                           preds   = c("x8", "x9", "x10"),
                           pm      = list(mar = 0.2, mcar = 0.1, mnar = 0.1),
                           snr     = list(mar = 5, mnar = 2.5),
                           pattern = "random")

missData$pattern
colMeans(is.na(missData$data))


missData <- imposeMissData(data     = testData,
                           targets  = list(mar  = c("y", "x1"),
                                           mcar = c("x2", "x3"),
                                           mnar = c("x4", "x5")
                                           ),
                           preds   = c("x8", "x9", "x10"),
                           pm      = 0.2,
                           snr     = 2.5,
                           pattern = "random")

colMeans(is.na(missData$data))

missData <- imposeMissData(data     = testData,
                           targets  = list(mar  = c("y", "x1"),
                                           mcar = c("x2", "x3"),
                                           mnar = c("x4", "x5")
                                           ),
                           preds   = c("x8", "x9", "x10"),
                           pm      = 0.2,
                           snr     = 2.5,
                           pattern = c(y  = "low",
                                       x5 = "random",
                                       x1 = "center",
                                       x4 = "tails")
                           )

colMeans(is.na(missData$data))

### plotImps ###

data(testImps)

plotImps(impList = testImps$impList,
         rMat    = testImps$rMat,
         typeVec = testImps$typeVec,
         interactive = TRUE)

### f2n ###

## Generate a factor:
x1 <- factor(c(0 : 5))
x1

## Naive type-casting:
as.numeric(x1)

## Safe type-casting:
f2n(x1)

### calcMode ###

## Find mode of an integer vector:
x1 <- sample(c(1 : 5), 100, TRUE)
calcMode(x1)

## Find mode of a factor:
x2 <- factor(sample(c(1 : 5), 100, TRUE))
calcMode(x2)

## Find mode of a character vector:
x3 <- sample(letters[1 : 5], 100, TRUE)
calcMode(x3)

## Find the mode of a continuous variable:
x4 <- rnorm(1000, 5, 2.5)
calcMode(x4, discrete = FALSE)

### rangeNorm ###

x0 <- runif(100, -3, 5)
x1 <- rangeNorm(x = x0)

range(x1)


source/SURF/R/plotImps.R
### Title:    Create Imputation Diagnostic Plots
### Author:   Kyle M. Lang
### Created:  2015-OCT-02
### Modified: 2018-JUN-01

##-------------------- COPYRIGHT & LICENSING INFORMATION ---------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of SURF.                                                ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program.  If not, see <http://www.gnu.org/licenses/>.           ##
##----------------------------------------------------------------------------##


plotImps <- function(impList,
                     rMat,
                     typeVec,
                     targetVar   = NULL,
                     interactive = FALSE)
{
    nImps          <- length(impList)
    missData       <- impList[[1]]
    missData[rMat] <- NA

    if(is.null(targetVar)) {
        levVec <- unlist(
            lapply(missData, function(x) length(unique(na.omit(x))))
            )
        indexVec <- which(colSums(rMat) > 0 & levVec > 1 & typeVec != "drop")
    } else if(is.character(targetVar)) {
        indexVec <- which(colnames(missData) %in% targetVar)
    } else {
        indexVec <- targetVar
    }

    for(i in indexVec) {
        target     <- missData[ , i]
        targetName <- colnames(missData)[i]
        catTarget  <- typeVec[i] %in% "cat"

        if(catTarget)
            imps <- lapply(impList,
                           function(x, rMat, index)
                               x[rMat[ , index], index],
                               rMat  = rMat,
                           index = i)
        else
            imps <- lapply(impList,
                           function(x, rMat, index) x[rMat[ , index], index],
                           rMat  = rMat,
                           index = i)

        if(catTarget) {
            hMat <- rbind(table(target), do.call("rbind", lapply(imps, table)))
            yMax <- max(hMat, na.rm = TRUE)

            barplot(height = hMat,
                    beside = TRUE,
                    col    = c("blue", rep("red", nImps)),
                    border = c("blue", rep("red", nImps)),
                    ylim   = c(0, yMax),
                    space  = c(0, nImps * 0.1),
                    width  = c(nImps, rep(1, nImps)),
                    main   =
                        paste0("Imputed (Red) vs. Observed (Blue) Values\nof ",
                               targetName),
                    ylab   = "Frequency",
                    xlab   = paste0("Level of ", targetName)
                    )
        } else {
            dens1 <- density(target, na.rm = TRUE)
            if(sum(rMat[ , i]) > 1) {
                dens2       <- lapply(imps, density)
                dens2RangeY <- range(lapply(dens2, function(x){x$y}))
                dens2RangeX <- range(lapply(dens2, function(x){x$x}))
            } else {
                dens2       <- imps
                dens2RangeY <- 0
                dens2RangeX <- range(unlist(imps))
            }
            yRange  <- range(dens1$y, dens2RangeY, na.rm = TRUE)
            xRange  <- range(dens1$x, dens2RangeX, na.rm = TRUE)

            plot(NULL,
                 main =
                     paste0("Densities of Imputations (Red) vs.\n",
                            "Observed Data (Blue) for ",
                            targetName),
                 ylab = "Density",
                 xlab = paste0("Value of ", targetName),
                 ylim = yRange,
                 xlim = xRange
                 )
            if(sum(rMat[ , i]) > 1)
                for(m in 1 : length(impList)) lines(dens2[[m]], col = "red")
            else
                for(m in 1 : length(impList))
                    points(y = 0, x = dens2[[m]], col = "red")

            lines(dens1, lwd = 3, col = "blue")
        }
        if(interactive & length(indexVec) > 1)
            readline("Hit any key to generate the next plot... ")
    }
}# END plotImps()



source/SURF/R/infoFunctions.R
### Title:    License/Warranty Information Functions for SURF
### Author:   Kyle M. Lang
### Created:  2016-MAY-09
### Modified: 2018-JUN-01

##-------------------- COPYRIGHT & LICENSING INFORMATION ---------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of SURF.                                                ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program.  If not, see <http://www.gnu.org/licenses/>.           ##
##----------------------------------------------------------------------------##

## Print warranty information:
surfW <- function() {
    lic <- readLines(system.file("LICENSE", package = "SURF"))
    
    start <- grep("15. Disclaimer of Warranty", lic)
    end   <- grep("END OF TERMS AND CONDITIONS", lic) - 1
    
    writeLines(lic[start : end])
}
                    
## Print license:
surfL <- function()
    writeLines(readLines(system.file("LICENSE", package = "SURF")))


source/SURF/R/supportFunctions.R
### Title:    Hidden support Functions for the SURF Package
### Author:   Kyle M. Lang
### Created:  2017-NOV-17
### Modified: 2018-JUN-01

##-------------------- COPYRIGHT & LICENSING INFORMATION ---------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of SURF.                                                ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program.  If not, see <http://www.gnu.org/licenses/>.           ##
##----------------------------------------------------------------------------##


## Print startup message:
.onAttach <- function(libname, pkgname) {
    version <- read.dcf(file   = system.file("DESCRIPTION", package = pkgname),
                        fields = "Version")
    
    greet <-
        strwrap(
            paste0("Loading: ",
                   pkgname,
                   " ",
                   version,
                   ", Copyright (C) ",
                   format(Sys.time(), "%Y"),
                   " Kyle M. Lang. ",
                   pkgname,
                   " is distributed under Version 3 of the GNU General Public License (GPL-3); execute 'surfL()' for details. ",
                   pkgname,
                   " comes with ABSOLUTELY NO WARRANTY; execute 'surfW()' for details. ",
                   pkgname,
                   " is beta software. Please report any bugs."),
            width = 81)
    
    for(i in greet) packageStartupMessage(i)
}


source/SURF/R/utilityFunctions.R
### Title:    Utility Functions for the SURF Package
### Author:   Kyle M. Lang
### Created:  2017-NOV-09
### Modified: 2018-JUN-01

##-------------------- COPYRIGHT & LICENSING INFORMATION ---------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of SURF.                                                ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program.  If not, see <http://www.gnu.org/licenses/>.           ##
##----------------------------------------------------------------------------##

## Range normalize a vector:
rangeNorm <- function(x,
                      oldMin = min(x),
                      oldMax = max(x),
                      newMin = 0.0,
                      newMax = 1.0)
{

    r0 <- oldMax - oldMin
    r1 <- newMax - newMin

    (x * r1 - min(x) * r1) / r0 + newMin
}

## Compute/estimate the mode:
calcMode <- function(x, discrete = TRUE) {
    if(discrete) {
        ## Find the modal value(s):
        tab  <- table(x, exclude = NULL)
        mode <- names(tab)[which.max(tab)]
        
        ## Break ties randomly:
        if(length(mode) > 1) mode <- mode[sample(c(1 : length(mode)), 1)]
        
        ## Cast output as numeric, if possible:
        if(is.numeric(x)) out <- as.numeric(mode)
        else              out <- mode
    } else {
        dens <- density(x, na.rm = TRUE)
        out  <- dens$x[which.max(dens$y)]
    }
    out
}

## Safely convert a factor to a numeric vector:
f2n <- function(x) as.numeric(as.character(x))


source/SURF/R/simulationFunctions.R
### Title:    Data Simulation Functions for the SURF Package
### Author:   Kyle M. Lang
### Created:  2017-OCT-25
### Modified: 2018-OCT-15

##-------------------- COPYRIGHT & LICENSING INFORMATION ---------------------##
##  Copyright (C) 2018 Kyle M. Lang <k.m.lang@uvt.nl>                         ##
##                                                                            ##
##  This file is part of SURF.                                                ##
##                                                                            ##
##  This program is free software: you can redistribute it and/or modify it   ##
##  under the terms of the GNU General Public License as published by the     ##
##  Free Software Foundation, either version 3 of the License, or (at you     ##
##  option) any later version.                                                ##
##                                                                            ##
##  This program is distributed in the hope that it will be useful, but       ##
##  WITHOUT ANY WARRANTY; without even the implied warranty of                ##
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General ##
##  Public License for more details.                                          ##
##                                                                            ##
##  You should have received a copy of the GNU General Public License along   ##
##  with this program.  If not, see <http://www.gnu.org/licenses/>.           ##
##----------------------------------------------------------------------------##


cleanInputs <- function() {
    with(parent.frame(),
    {
        ## Check/modify the mean vector:
        if(length(means) == 1) means <- rep(means, nVars)
        else if(length(means) != nVars)
            stop("'means' must have length 1 or length 'nVars'")

        ## Check/modify the scales vector:
        if(length(scales) == 1) scales <- rep(scales, nVars)
        else if(length(scales) != nVars)
            stop("'scales' must have length 1 or length 'nVars'")

        ## Check/modify the covariance matrix:
        if(is.vector(sigma) & length(sigma) == 1) {
            ## Generate a covariance matrix from 'scales' and 'sigma':
            w1 <- matrix(scales, nVars, nVars)
            w2 <- matrix(scales, nVars, nVars, byrow = TRUE)
            
            maxCov <- w1 * w2
            
            sigma       <- maxCov * sigma
            diag(sigma) <- scales^2
        }
        else if(!is.matrix(sigma))
            stop("'sigma' must be a matrix or a length-one vector")
    })
}



simRegData <- function(nObs,
                       r2,
                       sigma,
                       beta,
                       nVars           = ncol(sigma),
                       means           = 0,
                       scales          = 1,
                       itemsPerPred    = 1,
                       predReliability = 0.8)
{
    cleanInputs()
    
    ## Simulate predictor data:
    X <- cbind(1, rmvnorm(nObs, means, sigma))

    ## Simulate the outcome:
    eta    <- X %*% beta
    sigmaY <- (var(eta) / r2) - var(eta)
    y      <- eta + rnorm(nObs, 0, sqrt(sigmaY))

    ## Generate a latent structure wherein each predictor is indicated by
    ## 'itemsPerPred' observed variables:
    if(itemsPerPred > 1) {
        nItems   <- nVars * itemsPerPred
        loadings <- matrix(0, nItems, nVars)

        ## Populate loading matrix:
        for(m in 1 : nVars) {
            for(n in 1 : itemsPerPred) {
                offset                  <- (m - 1) * itemsPerPred
                loadings[n + offset, m] <- sqrt(predReliability)
            }
        }

        ## Populate error variance matrix:
        theta <- diag(rep(1 - predReliability, nItems))

        ## Generate observed items:
        X <- X[ , -1] %*% t(loadings) + rmvnorm(nObs, rep(0, nItems), theta)
        
        outDat           <- data.frame(y, X)
        colnames(outDat) <- c("y", paste0("x", 1 : nItems))
    }
    else {
        outDat           <- data.frame(y, X[ , -1])
        colnames(outDat) <- c("y", paste0("x", 1 : nVars))
    }
    outDat
}# END simRegData()



simCovData <- function(nObs, sigma, nVars = ncol(sigma), means = 0, scales = 1)
{
    cleanInputs()
    
    ## Simulate the data:
    outData           <- as.data.frame(rmvnorm(nObs, means, sigma))
    colnames(outData) <- paste0("x", 1 : nVars)

    outData
}# END simCovData()



makeRVec <- function(linPred, pm, snr, pattern) {
    ## Generate a linear predictor of missingness:
    noise   <- sd(linPred) / snr
    linPred <- linPred + rnorm(length(linPred), 0, noise)

    ## Use a probit model to simulate response propensities:
    probs <- pnorm(scale(linPred))

    ## 'pattern' selects which part of the target's distribution is missing:
    if(pattern == "random")
        pattern <- sample(c("low", "high", "tails", "center"), size = 1)

    ## Generate the missignness indicator:
    rVec <- switch(pattern,
                   low    = probs <= pm,
                   high   = probs >= (1 - pm),
                   tails  = probs <= (pm / 2) | probs >= (1 - (pm / 2)),
                   center = probs >= (0.5 - (pm / 2)) &
                       probs <= (0.5 + (pm / 2)),
                   stop("Please provide a valid 'pattern'")
                   )
    list(rVec = rVec, pattern = pattern)
}# END makeRVec()



## Check for various kinds of trival list entries:
empty <- function(x)
    missing(x) || is.null(x) || is.na(x) || length(x) == 0 || x == ""



imposeMissData <- function(data, targets, preds, pm, snr, pattern = "random") {
    ## Exclude empty 'targets' slots:
    targets <- targets[!sapply(targets, empty)]
    
    ## Which mechanisms should be simulated?
    mechs           <- c("mar", "mcar", "mnar")
    mechFlag        <- mechs %in% names(targets)
    names(mechFlag) <- mechs

    if(length(pattern) == 1) {
        pattern        <- rep(pattern, length(c(targets$mar, targets$mnar)))
        names(pattern) <- c(targets$mar, targets$mnar)
    }
    else {
        check <- all(c(targets$mar, targets$mnar) %in% names(pattern))
        
        if(!check)
            stop("Some elements in 'c(targets$mar, targets$mnar)' are not named in 'pattern'")
    }
    
    ## Create a vector to hold patterns used:
    patOut        <- rep(NA, ncol(data))
    names(patOut) <- colnames(data)
    
    ## Impose MCAR missing:
    if(mechFlag["mcar"]) {
        if(is.list(pm)) pm1 <- pm$mcar
        else            pm1 <- pm
        
        rMat <- matrix(
            as.logical(
                rbinom(n    = prod(dim(data[ , targets$mcar])),
                       size = 1,
                       prob = pm1)
            ),
            ncol = length(targets$mcar)
        )
        data[ , targets$mcar][rMat] <- NA
        patOut[targets$mcar]        <- "mcar"
    }
    
    ## Impose MAR missing:
    if(mechFlag["mar"]) {
        if(length(preds) == 1) linPred <- data[ , preds]
        else                   linPred <- rowSums(data[ , preds], na.rm = TRUE)
        
        if(any(is.na(linPred)))
            stop("MAR predictor matrix cannot have any entirely missing rows.")

        if(is.list(pm)) pm1 <- pm$mar
        else            pm1 <- pm
        
        if(is.list(snr)) snr1 <- snr$mar
        else             snr1 <- snr
        
        for(i in targets$mar) {
            tmp <- makeRVec(linPred = linPred,
                            pm      = pm1,
                            snr     = snr1,
                            pattern = pattern[i])
            data[tmp$rVec, i] <- NA
            patOut[i]         <- tmp$pattern
        }
    }
    
    ## Impose MNAR missing:
    if(mechFlag["mnar"]) {
        if(is.list(pm)) pm1 <- pm$mar
        else            pm1 <- pm
        
        if(is.list(snr)) snr1 <- snr$mar
        else             snr1 <- snr

        for(i in targets$mnar) {
            tmp <- makeRVec(linPred = data[ , i],
                            pm      = pm1,
                            snr     = snr1,
                            pattern = pattern[i])
            data[tmp$rVec, i] <- NA
            patOut[i]         <- tmp$pattern
        }
    }
    list(data = data, pattern = patOut)
}# END imposeMissing()
